src/appleseed/renderer/kernel/lighting/bdpt/bdptlightingengine.cpp

```diff
@@ -83,37 +123,317 @@ namespace
         {
         }
 
+        Spectrum ComputeGeometryTerm(
+            const ShadingContext&   shading_context,
+            const ShadingPoint&     shading_point,
+            const BDPTVertex&       a,
+            const BDPTVertex&       b)
+        {
+            double g = 1.0;
+
+            const Vector3d v = b.m_position - a.m_position;
+
+            /// TODO:: the special care have to be taken for these dot products when it comes to volume
+            // compute unnormalied dot products
+            g *= std::max(-foundation::dot(v, b.m_geometric_normal), 0.0);
+            g *= std::max(foundation::dot(v, a.m_geometric_normal), 0.0);
+
+            Spectrum result(0.0);
+
+            // do the visibility test
+            if (g > 0.0)
+            {
+                // compute inverse dist2 and normalize 2 dot products
+                const double dist2 = foundation::square_norm(v);
+                g /= (dist2 * dist2);
+                shading_context.get_tracer().trace_between_simple(
+                    shading_context,
+                    a.m_position + (v * 1.0e-6),
+                    b.m_position,
+                    shading_point.get_ray().m_time,
+                    VisibilityFlags::ShadowRay,
+                    shading_point.get_ray().m_depth,
+                    result);
+            }
+
+            return result * (float)g;
+        }
+
+        /// TODO:: precompute path density using fwd_pdf and rev_pdf.
+        float ComputePathDensity(
+            std::vector<BDPTVertex>             &light_vertices,
```

Comment:
Move `&` to the type's side:

    std::vector<BDPTVertex>&             light_vertices,